src/compose/images.coffee

```diff
@@ -0,0 +1,301 @@
+
+	cleanupDatabase: =>
+		@_withImagesFromDockerAndDB (dockerImages, supervisedImages) =>
+			Promise.map supervisedImages, (image) =>
+				# If the supervisor was interrupted between fetching an image and storing its id,
+				# some entries in the db might need to have the dockerImageId populated
+				if !image.dockerImageId?
+					id = _.find(dockerImages, (dockerImage) => @_matchesTagOrDigest(image, dockerImage))?.Id
+					if id?
+						@db.models('image').update(dockerImageId: id).where(image)
+						.then ->
+							image.dockerImageId = id
+			.then =>
+				_.filter(supervisedImages, (image) => !@_isAvailableInDocker(image, dockerImages))
+		.then (imagesToRemove) =>
+			ids = _.map(imagesToRemove, 'id')
+			@db.models('image').del().whereIn('id', ids)
+
+	getStatus: =>
+		@getAvailable()
+		.map (image) ->
+			image.status = 'Downloaded'
+			image.downloadProgress = null
+			return image
+		.then (images) =>
+			status = _.clone(@volatileState)
+			for image in images
+				status[image.imageId] ?= image
+			return _.values(status)
+
+	_getImagesForCleanup: =>
+		images = []
+		Promise.join(
+			@docker.getRegistryAndName(constants.supervisorImage)
+			@db.models('image').select('dockerImageId')
+			.map((image) -> image.dockerImageId)
+			(supervisorImageInfo, usedImageIds) =>
+				isDangling = (image) ->
+					# Looks like dangling images show up with these weird RepoTags and RepoDigests sometimes
+					(_.isEmpty(image.RepoTags) or _.isEqual(image.RepoTags, [ '<none>:<none>' ])) and
+						(_.isEmpty(image.RepoDigests) or _.isEqual(image.RepoDigests, [ '<none>@<none>' ]))
+				@docker.listImages(digests: true)
+				.map (image) =>
+					# Cleanup should remove truly dangling images (i.e. dangling and with no digests)
+					if isDangling(image) and not (image.Id in usedImageIds)
+						images.push(image.Id)
+					else if !_.isEmpty(image.RepoTags)
+						# We also remove images from the supervisor repository with a different tag
+						Promise.map image.RepoTags, (repoTag) =>
+							@docker.getRegistryAndName(repoTag)
+							.then ({ imageName, tagName }) ->
+								if imageName == supervisorImageInfo.imageName and tagName != supervisorImageInfo.tagName
+									images.push(image.Id)
+		)
+		.then =>
+			toCleanup = _.filter _.uniq(images), (image) =>
+				!@imageCleanupFailures[image]? or Date.now() - @imageCleanupFailures[image] > constants.imageCleanupErrorIgnoreTimeout
+			#console.log(toCleanup)
+			return toCleanup
+	inspectByName: (imageName) =>
+		@docker.getImage(imageName).inspect()
+		.catch NotFoundError, (err) =>
+			digest = imageName.split('@')[1]
+			Promise.try =>
+				if digest?
+					@db.models('image').where('name', 'like', "%@#{digest}").select()
+				else
+					@db.models('image').where(name: imageName).select()
+			.then (imagesFromDB) =>
+				for image in imagesFromDB
+					if image.dockerImageId?
+						return @docker.getImage(image.dockerImageId).inspect()
+				throw err
```

Comment:
Not really, this can throw or return an image if it finds it, tapCatch would always throw. Unless I'm misunderstanding how you mean I should use it here.